File: examples/exo/cpp/src/device_info.cpp
#include "exo_example_support.hpp"

#include <cstdlib>
#include <exception>
#include <iostream>
#include <string>

namespace {

void print_info(const qnbot::ExoDeviceInfo& info) {
    std::cout << "product=" << static_cast<int>(info.product)
              << " protocol=" << static_cast<int>(info.protocol_version.major)
              << '.' << static_cast<int>(info.protocol_version.minor)
              << " firmware=" << static_cast<int>(info.firmware.platform) << '.'
              << static_cast<int>(info.firmware.revision) << '.'
              << static_cast<int>(info.firmware.feature) << '.'
              << static_cast<int>(info.firmware.build);
    if (info.serial_number) std::cout << " sn=" << *info.serial_number;
    if (info.hardware_name) std::cout << " name=" << *info.hardware_name;
    std::cout << '\n';

    const auto& components = info.components;
    std::cout << "components encoders="
              << static_cast<int>(components.encoders.available) << '/'
              << static_cast<int>(components.encoders.total_slots)
              << " handsets=" << static_cast<int>(components.handsets.available)
              << '/' << static_cast<int>(components.handsets.total_slots)
              << " imus=" << static_cast<int>(components.imus.available) << '/'
              << static_cast<int>(components.imus.total_slots) << '\n';

    const auto& capabilities = info.capabilities;
    std::cout << "capabilities telemetry=" << capabilities.telemetry
              << " encoders=" << capabilities.encoders
              << " handsets=" << capabilities.handsets
              << " imus=" << capabilities.imus
              << " haptics=" << capabilities.haptics
              << " wireless=" << capabilities.wireless << '\n';
}

} // namespace

int main(int argc, char** argv) {
    try {
        std::string port;
        std::string name = "primary";
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--port") {
                port = example::require_value(argc, argv, index, argument);
            } else if (argument == "--name") {
                name = example::require_value(argc, argv, index, argument);
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (port.empty()) throw std::invalid_argument("--port is required");

        qnbot::Sdk sdk(example::serial_config(port, name));
        auto exo = sdk.exo();
        print_info(exo.device(name).get_device_info());
        exo.close();
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: exo/python/src/device_info.py
from __future__ import annotations

import argparse

from qnbot_sdk import Sdk, SerialConnection
from qnbot_sdk.exo import ExoConfig, ExoDeviceInfo


def create_sdk(port: str, name: str) -> Sdk:
    return Sdk(devices=(ExoConfig(name=name, connection=SerialConnection(port=port)),))


def print_info(info: ExoDeviceInfo) -> None:
    print(
        f"product={info.product.name} "
        f"protocol={info.protocol_version.major}.{info.protocol_version.minor} "
        f"firmware={info.firmware} sn={info.serial_number} name={info.hardware_name}"
    )
    print(
        "components "
        f"encoders={info.components.encoders.available}/"
        f"{info.components.encoders.total_slots} "
        f"handsets={info.components.handsets.available}/"
        f"{info.components.handsets.total_slots} "
        f"imus={info.components.imus.available}/{info.components.imus.total_slots}"
    )
    capabilities = info.capabilities
    print(
        "capabilities "
        f"telemetry={capabilities.telemetry} encoders={capabilities.encoders} "
        f"handsets={capabilities.handsets} imus={capabilities.imus} "
        f"haptics={capabilities.haptics} wireless={capabilities.wireless}"
    )


def main() -> None:
    arguments = argparse.ArgumentParser(
        description="Read Exo device information before start"
    )
    arguments.add_argument("--port", required=True, help="Serial port for the device")
    arguments.add_argument("--name", default="primary", help="Logical device name")
    options = arguments.parse_args()

    sdk = create_sdk(options.port, options.name)
    exo = sdk.exo()
    device = exo.device(options.name)
    print_info(device.get_device_info())
    exo.close()


if __name__ == "__main__":
    main()
